ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala

```diff
@@ -0,0 +1,274 @@
+import com.digitalasset.ledger.api.v1.transaction.Transaction
+import com.digitalasset.ledger.api.v1.transaction_filter.{
+  Filters,
+  InclusiveFilters,
+  TransactionFilter
+}
+import com.digitalasset.ledger.api.v1.transaction_service.{
+  GetLedgerEndRequest,
+  GetTransactionsRequest,
+  GetTransactionsResponse,
+  TransactionServiceGrpc
+}
+import com.digitalasset.ledger.api.v1.value.{Identifier, Record, RecordField, Value}
+import com.digitalasset.platform.sandbox.SandboxServer
+import com.digitalasset.platform.sandbox.config.SandboxConfig
+import com.google.protobuf.timestamp.Timestamp
+import io.grpc.ManagedChannel
+import io.grpc.netty.NettyChannelBuilder
+import io.netty.channel.nio.NioEventLoopGroup
+import io.netty.util.concurrent.DefaultThreadFactory
+import org.slf4j.LoggerFactory
+
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Try}
+
+private[acceptance] sealed abstract case class LedgerSession(
+    configuration: LedgerSessionConfiguration) {
+
+  protected val channel: ManagedChannel
+
+  def close(): Unit
+
+  private lazy val commandService = CommandServiceGrpc.stub(channel)
+  private lazy val transactionService = TransactionServiceGrpc.stub(channel)
+  private lazy val ledgerIdentityService = LedgerIdentityServiceGrpc.stub(channel)
+  private lazy val partyManagementService = PartyManagementServiceGrpc.stub(channel)
+  private lazy val timeService = TimeServiceGrpc.stub(channel)
+
+  private[infrastructure] def ledgerId()(implicit context: ExecutionContext): Future[String] =
+    for {
+      response <- ledgerIdentityService.getLedgerIdentity(new GetLedgerIdentityRequest)
+    } yield response.ledgerId
+
+  private[infrastructure] def ledgerEnd()(
+      implicit context: ExecutionContext): Future[LedgerOffset] =
+    for {
+      id <- ledgerId
+      response <- transactionService.getLedgerEnd(new GetLedgerEndRequest(id))
+    } yield response.offset.get
+
+  private[infrastructure] def transactionFilter(party: String, templateIds: Seq[Identifier]) =
+    new TransactionFilter(Map(party -> filter(templateIds)))
+
+  private[infrastructure] def filter(templateIds: Seq[Identifier]): Filters =
+    new Filters(if (templateIds.isEmpty) None else Some(new InclusiveFilters(templateIds)))
+
+  private[infrastructure] def allocateParty()(implicit context: ExecutionContext): Future[String] =
+    partyManagementService.allocateParty(new AllocatePartyRequest()).map(_.partyDetails.get.party)
+
+  private[infrastructure] def allocateParties(n: Int)(
+      implicit context: ExecutionContext): Future[Vector[String]] =
+    Future.sequence(Vector.fill(n)(allocateParty()))
+
+  private[infrastructure] def transactionsUntilNow(
+      begin: LedgerOffset,
+      party: String,
+      templateIds: Identifier*)(implicit context: ExecutionContext): Future[Vector[Transaction]] =
+    for {
+      id <- ledgerId()
+      end <- ledgerEnd()
```

Comment:
Technically you don't need to fetch the ledger end first. You can use the [LedgerBoundary#LEDGER_END](https://github.com/digital-asset/daml/blob/master/ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/ledger_offset.proto#L31) instead and the ledger will do the "what's the current ledger end" check for you.

Reply:
Fixed by https://github.com/digital-asset/daml/pull/2282/commits/577409f56cb48d68561a79521e4abac9db92cc59